ce/api/streamflow/watershed_streams.py

```diff
@@ -0,0 +1,194 @@
+        Necessary because, depending on whether lon and lat dimensions
+        increase or decrease with increasing index, a move north or east is
+        represented by an offset of +1 or -1, respectively.
+        TODO: Compute this internally?
+    :param debug: Boolean indicating whether this function should compute
+        and return debug information.
+    :return: A list of lists of cells that drain into `target`.
+
+    Notes:
+
+    - In this function, a cell is represented by an (x, y) index pair.
+
+    - Routing graphs can and in practice do contain cycles. Variable `visited`
+    is used to determine whether a cell has already been visited during the
+    traversal of the routing graph, i.e., whether we are cycling, and if so
+    not to repeat that subgraph.
+
+    - Variable streams is a list of streams, where each stream is a list of
+    cells that drain into the target
+
+    - Variable 'first' is a boolean flag that is true when the current cell is
+    also the target cell. This is used in the case that the target cell has no
+    upstream neighbours
+
+    - Variable 'number_of_streams' counts the number of streams in the watershed. 
+    This variable is used to ensure the cells are inserted into the right stream
+    """
+    visited = set()
+    streams = []
+    number_of_streams = 0
+
+    def upstream(cell, first):
+        """Generates list of lists, each list being a stream of cells upstream of
+        the inputted cell. This list is named 'streams', a nonlocal variable.
+
+        - Variable 'eligible' is generated for each cell, and is a list of all
+        'eligible' neighbours. Eligible is defined as a neighbour that is
+        upstream of the current cell and that also has not been visited.
+        """
+        nonlocal visited
+        nonlocal number_of_streams
+        nonlocal streams
+
+        visited |= {cell}
+        eligible = [
+            neighbour
+            for neighbour in neighbours(cell)
+            if neighbour not in visited
+            and is_upstream(neighbour, cell, routing, direction_map)
+        ]
+
+        # If there are no eligible neighbours for the first cell, connection
+        # should be []. If there are eligible neighbours, the first stream
+        # must be started. Otherwise, the cell is a part of a stream and can
+        # be added to the current stream.
+        if eligible == [] and first:
+            return
+        elif first:
+            streams.append([cell])
+        else:
+            streams[number_of_streams].append(cell)
+
+        # At this point, the first cell is finished, so the flag is no longer
+        # true
+
+        # For each eligible neighbour, we want to add them to a stream and
+        # move upstream on those neighbours (call the function again)
+        for neighbour in eligible:
+            upstream(neighbour, False)
+            if neighbour != eligible[-1]:
```

Comment:
Add a little comment explaining why you don't want to start a new stream after the last item.